ompi/mca/osc/portals4/osc_portals4_comm.c

```diff
@@ -179,95 +179,1808 @@ ompi_osc_portals4_get_dt(struct ompi_datatype_t *dt, ptl_datatype_t *ptl_dt)
+
+    md.start = module->origin_iovec_list;
+    md.length = iovec_count;
+    md.options = PTL_IOVEC | PTL_MD_EVENT_SUCCESS_DISABLE | PTL_MD_EVENT_CT_REPLY | PTL_MD_EVENT_CT_ACK;
+    md.eq_handle = mca_osc_portals4_component.matching_eq_h;
+    md.ct_handle = module->ct_h;
+    ret = PtlMDBind(module->ni_h, &md, &module->origin_iovec_md_h);
+    if (PTL_OK != ret) {
+        opal_output_verbose(1, ompi_osc_base_framework.framework_output,
+                            "%s:%d: PtlMDBind(iovec) failed: %d\n",
+                            __FILE__, __LINE__, ret);
+        return ret;
+    }
+
+    OPAL_OUTPUT_VERBOSE((90, ompi_osc_base_framework.framework_output,
+                 "%s,%d Put(origin_count=%d, origin_lb=%lu, target_count=%d, target_lb=%lu, length=%lu, op_count=%ld)",
+                 __FUNCTION__, __LINE__, origin_count, origin_lb, target_count, target_lb, length, module->opcount));
+    ret = segmentedPut(&module->opcount,
+                       module->origin_iovec_md_h,
+                       (ptl_size_t) origin_lb,
+                       length,
+                       module->atomic_max,
+                       PTL_ACK_REQ,
+                       peer,
+                       module->pt_idx,
+                       module->match_bits,
+                       offset + target_lb,
+                       NULL,
+                       0);
+    if (OMPI_SUCCESS != ret) {
+        return ret;
+    }
+
+    return OMPI_SUCCESS;
+}
+
+/* perform atomic operation on iovec local and contiguous remote */
+static int
+atomic_from_iovec(ompi_osc_portals4_module_t *module,
+                  const void       *origin_address,
+                  int               origin_count,
+                  ompi_datatype_t  *origin_datatype,
+                  ptl_process_t     peer,
+                  int               target_count,
+                  ompi_datatype_t  *target_datatype,
+                  size_t            offset,
+                  ptl_pt_index_t    pt_index,
+                  ptl_match_bits_t  match_bits,
+                  struct ompi_op_t *op,
+                  void             *user_ptr)
+{
+    int ret;
+    size_t size;
+    OPAL_PTRDIFF_TYPE length, origin_lb, target_lb, extent;
+    ptl_md_t md;
+    ptl_op_t ptl_op;
+    ptl_datatype_t ptl_dt;
+
+    if (module->origin_iovec_md_h != PTL_INVALID_HANDLE) {
+        PtlMDRelease(module->origin_iovec_md_h);
+        free(module->origin_iovec_list);
+        module->origin_iovec_md_h = PTL_INVALID_HANDLE;
+        module->origin_iovec_list = NULL;
+    }
+
+    ptl_size_t iovec_count=0;
+    create_iov_list(
+        origin_address,
+        origin_count,
+        origin_datatype,
+        &module->origin_iovec_list,
+        &iovec_count);
+
+    ret = ompi_osc_portals4_get_dt(target_datatype, &ptl_dt);
+    if (OMPI_SUCCESS != ret) {
+        opal_output(ompi_osc_base_framework.framework_output,
+                "MPI_Accumulate: datatype is not currently supported");
```

Comment:
technically you don't know it's an accumulate op at this point for the opal output...